In [1]:
TOPIC_WEIBO_PATH = r'..\data\crawler\topic_weibo.parquet'
TOPIC_COMMENT_PATH = r'..\data\crawler\topic_comment.parquet'
USER_INFO_PATH = r'..\data\crawler\user_info.parquet'
USER_WEIBO_PATH = r'..\data\crawler\user_weibo.parquet'

In [2]:
import pandas as pd

df_topic_weibo = pd.read_parquet(TOPIC_WEIBO_PATH)
df_topic_comment = pd.read_parquet(TOPIC_COMMENT_PATH)
df_user_info = pd.read_parquet(USER_INFO_PATH)
df_user_weibo = pd.read_parquet(USER_WEIBO_PATH)

# 📊 数据探索

先全面了解四个数据表的结构、字段类型、缺失情况和基本统计特征，为后续清洗策略提供依据。

In [3]:
print("=" * 60)
print("📋 df_topic_weibo（话题微博）")
print(f"Shape: {df_topic_weibo.shape}")
print(f"\nColumns: {df_topic_weibo.columns.tolist()}")
print(f"\nDtypes:\n{df_topic_weibo.dtypes}")
print(f"\n缺失值:\n{df_topic_weibo.isnull().sum()}")
print(f"\n前3行:")
df_topic_weibo.head(3)

📋 df_topic_weibo（话题微博）
Shape: (4747, 11)

Columns: ['weibo_id', 'content', 'create_time', 'create_time_ts', 'like_count', 'comment_count', 'repost_count', 'user_id', 'screen_name', 'gender', 'topic']

Dtypes:
weibo_id                   int64
content                   object
create_time       datetime64[ns]
create_time_ts             int64
like_count                 int64
comment_count              int64
repost_count               int64
user_id                    int64
screen_name               object
gender                    object
topic                     object
dtype: object

缺失值:
weibo_id          0
content           0
create_time       0
create_time_ts    0
like_count        0
comment_count     0
repost_count      0
user_id           0
screen_name       0
gender            0
topic             0
dtype: int64

前3行:


,weibo_id,content,create_time,create_time_ts,like_count,comment_count,repost_count,user_id,screen_name,gender,topic
0,5206726806867173,【#习近平九三重要讲话#】九三胜利日，习近平总书记重要讲话金句来了！#中华民族伟大复兴势不可...,2025-09-03 09:44:30,1756892670,129,22,36,2442273073,内蒙古青年,m,中华民族伟大复兴势不可挡
1,5199292492154667,【#大理苍山走失8岁男童不幸遇难#，涉事夏令营机构负责人正接受调查】近日，一自闭症男孩在大理...,2025-08-13 21:23:11,1755120191,2777,1230,75,1744332207,钱江视频,m,大理苍山失踪8岁男童不幸遇难
2,5119182115184000,#张云雷0111生日快乐##张云雷一岁一礼#2023.01.02广州佛山流行金曲嘉年华，主办...,2025-01-04 19:53:09,1736020389,192,113,13,7164946878,典型摩羯座月亮,f,2025继续Passion


In [4]:
print("=" * 60)
print("📋 df_topic_comment（话题评论）")
print(f"Shape: {df_topic_comment.shape}")
print(f"\nColumns: {df_topic_comment.columns.tolist()}")
print(f"\nDtypes:\n{df_topic_comment.dtypes}")
print(f"\n缺失值:\n{df_topic_comment.isnull().sum()}")
print(f"\n前3行:")
df_topic_comment.head(3)

📋 df_topic_comment（话题评论）
Shape: (121807, 12)

Columns: ['comment_id', 'parent_id', 'weibo_id', 'user_id', 'screen_name', 'content', 'create_time', 'create_time_ts', 'sub_comment_count', 'like_count', 'ip_location', 'gender']

Dtypes:
comment_id                    int64
parent_id                     int64
weibo_id                      int64
user_id                       int64
screen_name                  object
content                      object
create_time          datetime64[ns]
create_time_ts                int64
sub_comment_count             int64
like_count                    int64
ip_location                  object
gender                       object
dtype: object

缺失值:
comment_id           0
parent_id            0
weibo_id             0
user_id              0
screen_name          0
content              0
create_time          0
create_time_ts       0
sub_comment_count    0
like_count           0
ip_location          0
gender               0
dtype: int64

前3行:


,comment_id,parent_id,weibo_id,user_id,screen_name,content,create_time,create_time_ts,sub_comment_count,like_count,ip_location,gender
0,5119246901709277,-1,5119182115184000,7164946878,典型摩羯座月亮,逐光,2025-01-05 00:10:35,1736035835,0,3,山东,f
1,5119305493250577,-1,5119182115184000,5119341911,柔风细雨润无声,张云雷 世界上所有的惊喜和好运，都是你累积的温柔和善良。@小辫儿张云雷,2025-01-05 04:03:24,1736049804,0,7,天津,f
2,5119203016705315,-1,5119182115184000,6458485512,柠檬水萌不萌,张云雷 世界上所有的惊喜和好运，都是你累积的温柔和善良。@小辫儿张云雷,2025-01-04 21:16:11,1736025371,0,5,江苏,f


In [5]:
print("=" * 60)
print("📋 df_user_info（用户信息）")
print(f"Shape: {df_user_info.shape}")
print(f"\nColumns: {df_user_info.columns.tolist()}")
print(f"\nDtypes:\n{df_user_info.dtypes}")
print(f"\n缺失值:\n{df_user_info.isnull().sum()}")
print(f"\n前3行:")
df_user_info.head(3)

📋 df_user_info（用户信息）
Shape: (11012, 13)

Columns: ['user_id', 'screen_name', 'gender', 'ip_location', 'registration_time', 'total_weibo_count', 'follower_count', 'following_count', 'description', 'verified', 'verified_type', 'verified_type_name', 'user_rank']

Dtypes:
user_id               object
screen_name           object
gender                object
ip_location           object
registration_time     object
total_weibo_count      int64
follower_count         int64
following_count        int64
description           object
verified                bool
verified_type          int64
verified_type_name    object
user_rank              int64
dtype: object

缺失值:
user_id               0
screen_name           0
gender                0
ip_location           0
registration_time     0
total_weibo_count     0
follower_count        0
following_count       0
description           0
verified              0
verified_type         0
verified_type_name    0
user_rank             0
dtype: int64

前3行:


,user_id,screen_name,gender,ip_location,registration_time,total_weibo_count,follower_count,following_count,description,verified,verified_type,verified_type_name,user_rank
0,1000123015,光打雷不闪电,m,江苏,2012-02-03,8411,1182,2209,顺势而为！,False,-1,普通用户,23
1,1003072032,晨柳1207,f,河南,2014-02-28,4269,33000,4317,我频繁记录着，只因为生活值得,True,0,个人认证,4
2,1003160191,大糊小狐,m,未知,2014-02-28,7681,3058,439,得天道者得人心，得人心者得天下。入世出世皆寂寞，豪门寒门终归尘。微头条做得不错，没啥限制，特...,False,-1,普通用户,32


In [6]:
print("=" * 60)
print("📋 df_user_weibo（用户历史微博）")
print(f"Shape: {df_user_weibo.shape}")
print(f"\nColumns: {df_user_weibo.columns.tolist()}")
print(f"\nDtypes:\n{df_user_weibo.dtypes}")
print(f"\n缺失值:\n{df_user_weibo.isnull().sum()}")
print(f"\n前3行:")
df_user_weibo.head(3)

📋 df_user_weibo（用户历史微博）
Shape: (919646, 12)

Columns: ['weibo_id', 'user_id', 'screen_name', 'content', 'create_time', 'create_time_ts', 'like_count', 'comment_count', 'repost_count', 'topics', 'at_users', 'reposted_weibo_id']

Dtypes:
weibo_id                      int64
user_id                       int64
screen_name                  object
content                      object
create_time          datetime64[ns]
create_time_ts                int64
like_count                    int64
comment_count                 int64
repost_count                  int64
topics                       object
at_users                     object
reposted_weibo_id             int64
dtype: object

缺失值:
weibo_id             0
user_id              0
screen_name          0
content              0
create_time          0
create_time_ts       0
like_count           0
comment_count        0
repost_count         0
topics               0
at_users             0
reposted_weibo_id    0
dtype: int64

前3行:


,weibo_id,user_id,screen_name,content,create_time,create_time_ts,like_count,comment_count,repost_count,topics,at_users,reposted_weibo_id
0,5271091661510186,1000123015,光打雷不闪电,转发微博,2026-02-28 00:27:27,1772209647,0,0,0,,,5271019791061916
1,5271019791061916,7782884695,南归孤夷,这是一位外国博主的科普，他讲述了中国华工在美国的血泪史，国内很少有人提及#美国发动网攻侵占全...,2026-02-27 19:41:52,1772192512,82,5,18,"美国发动网攻侵占全球虚拟资产,海外新鲜事,热点现场",,-1
2,5270988024973765,1000123015,光打雷不闪电,转发微博,2026-02-27 17:35:38,1772184938,1,4,1,,,5270947783510467


In [7]:
# 深入探索文本和行为特征
print("=" * 60)
print("🔍 深入数据特征分析")
print("=" * 60)

# 1. 各表重复记录
print("\n【重复记录检查】")
print(f"  topic_weibo  重复 weibo_id: {df_topic_weibo['weibo_id'].duplicated().sum()}")
print(f"  topic_comment 重复 comment_id: {df_topic_comment['comment_id'].duplicated().sum()}")
print(f"  user_info     重复 user_id: {df_user_info['user_id'].duplicated().sum()}")
print(f"  user_weibo    重复 weibo_id: {df_user_weibo['weibo_id'].duplicated().sum()}")

# 2. 文本长度分布
print("\n【文本长度统计 (字符数)】")
for name, df, col in [
    ("topic_weibo", df_topic_weibo, "content"),
    ("topic_comment", df_topic_comment, "content"),
    ("user_weibo", df_user_weibo, "content"),
]:
    lengths = df[col].str.len()
    print(f"  {name}: mean={lengths.mean():.1f}, median={lengths.median():.1f}, "
          f"min={lengths.min()}, max={lengths.max()}")

# 3. 空或极短文本
print("\n【极短文本 (<=2字符)】")
for name, df, col in [
    ("topic_weibo", df_topic_weibo, "content"),
    ("topic_comment", df_topic_comment, "content"),
    ("user_weibo", df_user_weibo, "content"),
]:
    short_count = (df[col].str.len() <= 2).sum()
    print(f"  {name}: {short_count} 条 ({short_count/len(df)*100:.2f}%)")

# 4. 转发微博特征（user_weibo）
print("\n【user_weibo 转发分析】")
repost_mask = df_user_weibo["reposted_weibo_id"] != -1
print(f"  转发微博: {repost_mask.sum()} 条 ({repost_mask.sum()/len(df_user_weibo)*100:.1f}%)")
print(f"  原创微博: {(~repost_mask).sum()} 条 ({(~repost_mask).sum()/len(df_user_weibo)*100:.1f}%)")

# 5. "转发微博" 纯文本
pure_repost = (df_user_weibo["content"] == "转发微博").sum()
print(f"  内容为'转发微博': {pure_repost} 条 ({pure_repost/len(df_user_weibo)*100:.1f}%)")

# 6. 话题分布
print("\n【topic_weibo 话题分布】")
print(df_topic_weibo["topic"].value_counts().head(10))

# 7. 性别分布
print("\n【性别分布】")
for name, df in [("topic_weibo", df_topic_weibo), ("topic_comment", df_topic_comment),
                  ("user_info", df_user_info)]:
    print(f"  {name}: {df['gender'].value_counts().to_dict()}")

# 8. user_info 中 user_id 类型
print(f"\n【user_info.user_id 类型: {df_user_info['user_id'].dtype}】")
print(f"  样例: {df_user_info['user_id'].head(3).tolist()}")

🔍 深入数据特征分析

【重复记录检查】
  topic_weibo  重复 weibo_id: 0
  topic_comment 重复 comment_id: 0
  user_info     重复 user_id: 0
  user_weibo    重复 weibo_id: 321450

【文本长度统计 (字符数)】
  topic_weibo: mean=207.7, median=109.0, min=2, max=4649
  topic_comment: mean=17.6, median=11.0, min=0, max=263
  user_weibo: mean=133.1, median=54.0, min=0, max=8873

【极短文本 (<=2字符)】
  topic_weibo: 1 条 (0.02%)
  topic_comment: 12912 条 (10.60%)
  user_weibo: 18194 条 (1.98%)

【user_weibo 转发分析】
  转发微博: 220362 条 (24.0%)
  原创微博: 699284 条 (76.0%)
  内容为'转发微博': 73974 条 (8.0%)

【topic_weibo 话题分布】
topic
                    197
跨晚新疆美女采销             17
律师解读吸毒记录封存不是彻底清除     17
赵露思艾特银河酷娱            16
官方通报女司机亮证逼迫让路事件      16
湾区升明月                15
西贝承认部分菜品是隔夜菜         15
李刚受贿1.02亿余元          15
宗馥莉已经辞职              15
19岁女大学生景区内遭捅多刀身亡     15
Name: count, dtype: int64

【性别分布】
  topic_weibo: {'m': 3360, 'f': 1387}
  topic_comment: {'f': 62377, 'm': 59430}
  user_info: {'f': 5908, 'm': 5104}

【user_info.user_id 类型: object】
  样例: ['100012

In [8]:
import re

# 微博文本中的特殊内容特征分析
print("=" * 60)
print("🔍 文本内容特征分析（微博特有元素）")
print("=" * 60)

sample_texts = df_topic_weibo["content"].head(20).tolist() + df_user_weibo["content"].head(20).tolist()

# 检测常见模式
patterns = {
    "URL链接": r'https?://\S+',
    "话题标签 #xxx#": r'#[^#]+#',
    "@用户": r'@[\w\u4e00-\u9fff]+',
    "表情 [xxx]": r'\[[^\[\]]+\]',
    "HTML标签": r'<[^>]+>',
    "转发标记": r'转发微博',
    "回复标记": r'回复@',
}

for table_name, df in [("topic_weibo", df_topic_weibo), ("topic_comment", df_topic_comment),
                        ("user_weibo", df_user_weibo)]:
    print(f"\n--- {table_name} (共{len(df)}条) ---")
    for pat_name, pat in patterns.items():
        count = df["content"].str.contains(pat, regex=True, na=False).sum()
        print(f"  {pat_name}: {count} 条 ({count/len(df)*100:.1f}%)")

# 查看 topic_comment 中空文本情况
print("\n--- topic_comment 空文本内容 ---")
empty_comments = df_topic_comment[df_topic_comment["content"].str.len() == 0]
print(f"完全空文本: {len(empty_comments)} 条")

# user_weibo 中内容为空的情况
print("\n--- user_weibo 空文本内容 ---")
empty_weibo = df_user_weibo[df_user_weibo["content"].str.len() == 0]
print(f"完全空文本: {len(empty_weibo)} 条")

# 查看 topic_comment 中极短文本样例
print("\n--- topic_comment 极短文本样例 (<=2字符) ---")
short_comments = df_topic_comment[df_topic_comment["content"].str.len() <= 2]["content"].value_counts().head(15)
print(short_comments)

🔍 文本内容特征分析（微博特有元素）

--- topic_weibo (共4747条) ---
  URL链接: 0 条 (0.0%)
  话题标签 #xxx#: 4687 条 (98.7%)
  @用户: 350 条 (7.4%)
  表情 [xxx]: 9 条 (0.2%)
  HTML标签: 0 条 (0.0%)
  转发标记: 1 条 (0.0%)
  回复标记: 0 条 (0.0%)

--- topic_comment (共121807条) ---
  URL链接: 0 条 (0.0%)
  话题标签 #xxx#: 1963 条 (1.6%)
  @用户: 6694 条 (5.5%)
  表情 [xxx]: 40 条 (0.0%)
  HTML标签: 0 条 (0.0%)
  转发标记: 647 条 (0.5%)
  回复标记: 3409 条 (2.8%)

--- user_weibo (共919646条) ---
  URL链接: 7 条 (0.0%)
  话题标签 #xxx#: 437482 条 (47.6%)
  @用户: 140876 条 (15.3%)
  表情 [xxx]: 2789 条 (0.3%)
  HTML标签: 972 条 (0.1%)
  转发标记: 76949 条 (8.4%)
  回复标记: 2522 条 (0.3%)

--- topic_comment 空文本内容 ---
完全空文本: 6517 条

--- user_weibo 空文本内容 ---
完全空文本: 9302 条

--- topic_comment 极短文本样例 (<=2字符) ---
content
      6517
关注     257
晚安     254
是的     203
加油     140
早安     133
支持     121
接      103
期待      94
是啊      81
致敬      71
厉害      68
1       68
唉       64
好       59
Name: count, dtype: int64


In [9]:
# 重复 weibo_id 分析 & 数据关系完整性
print("=" * 60)
print("🔍 user_weibo 重复分析 & 表间关系")
print("=" * 60)

# user_weibo 重复 weibo_id 是什么情况？
dup_ids = df_user_weibo[df_user_weibo["weibo_id"].duplicated(keep=False)]
print(f"\n重复 weibo_id 涉及 {dup_ids['weibo_id'].nunique()} 个唯一 weibo_id，共 {len(dup_ids)} 行")
# 是同一用户的重复还是不同用户的？
dup_sample = dup_ids.groupby("weibo_id").agg(
    n_users=("user_id", "nunique"),
    n_rows=("user_id", "count")
).reset_index()
print(f"多用户重复（转发同一条）: {(dup_sample['n_users'] > 1).sum()}")
print(f"同用户重复（完全重复）: {(dup_sample['n_users'] == 1).sum()}")

# 表间关系
print("\n--- 表间关联分析 ---")
# topic_comment 的 weibo_id 是否都在 topic_weibo 中
comment_weibo_ids = set(df_topic_comment["weibo_id"].unique())
topic_weibo_ids = set(df_topic_weibo["weibo_id"].unique())
print(f"topic_comment 中 weibo_id 能关联到 topic_weibo: "
      f"{len(comment_weibo_ids & topic_weibo_ids)}/{len(comment_weibo_ids)}")

# user_info 与 topic_weibo/topic_comment 的用户覆盖
user_info_ids = set(df_user_info["user_id"].astype(str).unique())
topic_weibo_user_ids = set(df_topic_weibo["user_id"].astype(str).unique())
topic_comment_user_ids = set(df_topic_comment["user_id"].astype(str).unique())
user_weibo_user_ids = set(df_user_weibo["user_id"].astype(str).unique())

print(f"user_info 覆盖 topic_weibo 用户: "
      f"{len(user_info_ids & topic_weibo_user_ids)}/{len(topic_weibo_user_ids)}")
print(f"user_info 覆盖 topic_comment 用户: "
      f"{len(user_info_ids & topic_comment_user_ids)}/{len(topic_comment_user_ids)}")
print(f"user_weibo 覆盖 user_info 用户: "
      f"{len(user_weibo_user_ids & user_info_ids)}/{len(user_info_ids)}")

# user_info 中有多少用户在 user_weibo 中有微博？
print(f"\n--- user_weibo 每用户微博数量分布 ---")
print(df_user_weibo.groupby("user_id").size().describe())

🔍 user_weibo 重复分析 & 表间关系

重复 weibo_id 涉及 68891 个唯一 weibo_id，共 390341 行
多用户重复（转发同一条）: 82
同用户重复（完全重复）: 68809

--- 表间关联分析 ---
topic_comment 中 weibo_id 能关联到 topic_weibo: 4704/4704
user_info 覆盖 topic_weibo 用户: 373/2116
user_info 覆盖 topic_comment 用户: 11012/83606
user_weibo 覆盖 user_info 用户: 11011/11012

--- user_weibo 每用户微博数量分布 ---
count    46008.000000
mean        19.988828
std        123.687056
min          1.000000
25%          1.000000
50%          4.000000
75%         29.000000
max      21429.000000
dtype: float64


# 🧹 数据清洗

## 清洗策略概述

基于以上数据探索，制定以下清洗策略：

### 1. 全局清洗
- **去重**：`user_weibo` 存在大量同用户重复 weibo_id（68,809 组），需去重
- **类型统一**：`user_info.user_id` 为 `object` 类型，其他表为 `int64`，需统一为 `int64` 以便关联

### 2. 文本清洗（保留情绪信号）
- **移除 HTML 标签**：`user_weibo` 中 972 条含 HTML 标签
- **移除 URL 链接**
- **清洗话题标签 `#xxx#`**：提取标签内容保留，去掉 `#` 符号（标签本身含情绪信息）
- **保留 `@用户` 引用**：体现社交互动关系，仅在分析文本时按需移除
- **保留表情符号 `[xxx]`**：微博表情是重要的情绪信号
- **处理 `回复@xxx：` 格式**：提取回复目标用户信息后清理前缀

### 3. 行为数据清洗
- **转发微博处理**：内容为"转发微博"的无文本信息价值，但保留 `reposted_weibo_id` 作为转发关系信号，添加 `is_repost` 标记列
- **空文本处理**：
  - `topic_comment`：6,517 条空评论 → 移除（无情绪信息）
  - `user_weibo`：9,302 条空微博 → 移除

### 4. 数据增强
- **添加文本长度特征**：`text_length` 列
- **添加互动量特征**：`engagement = like_count + comment_count + repost_count`
- **添加时间特征**：年/月/日/时/星期（已有代码）
- **添加转发标记**：`is_repost` 列
- **用户活跃度统计**：每位用户的微博数量

### 5. 数据质量过滤
- **过滤极短评论**（<=2字符且无情绪价值的，如空字符串）
- **保留极短但有情绪表达的评论**（如"加油""唉""晚安"等，这些本身就是情绪表达）

## Step 1: 重新加载原始数据 & 去重 & 类型统一

In [10]:
import pandas as pd
import re

# ========== 重新加载原始数据 ==========
df_topic_weibo = pd.read_parquet(TOPIC_WEIBO_PATH)
df_topic_comment = pd.read_parquet(TOPIC_COMMENT_PATH)
df_user_info = pd.read_parquet(USER_INFO_PATH)
df_user_weibo = pd.read_parquet(USER_WEIBO_PATH)


# ========== 1.1 去重 ==========
# topic_weibo: 无重复，跳过
# topic_comment: 无重复，跳过
# user_info: 无重复，跳过
# user_weibo: 68,809 组同用户重复 weibo_id
# 策略：对于同一 weibo_id，保留第一条（同用户重复取其一）；
#        对于多用户同 weibo_id（82 组，转发关系），按 user_id 区分后保留
df_user_weibo = df_user_weibo.drop_duplicates(subset=["weibo_id", "user_id"], keep="first")
print(f"\n✅ user_weibo 去重后: {len(df_user_weibo):>10,}")

# ========== 1.2 统一 user_id 类型为 int64 ==========
df_user_info["user_id"] = df_user_info["user_id"].astype("int64")
print(f"✅ user_info.user_id 类型统一为: {df_user_info['user_id'].dtype}")


# 记录各表原始行数，用于全程追踪保留率
_ORIGINAL_COUNTS = {
    "topic_weibo": len(df_topic_weibo),
    "topic_comment": len(df_topic_comment),
    "user_info": len(df_user_info),
    "user_weibo": len(df_user_weibo),
}

def retention_rate(table: str, current_df) -> str:
    """计算相对于原始数据的保留率。"""
    orig = _ORIGINAL_COUNTS[table]
    cur = len(current_df)
    return f"{cur:,} / {orig:,} ({cur / orig * 100:.1f}% 保留)"

print("\n📦 原始数据量:")
print(f"  topic_weibo:   {len(df_topic_weibo):>10,}")
print(f"  topic_comment: {len(df_topic_comment):>10,}")
print(f"  user_info:     {len(df_user_info):>10,}")
print(f"  user_weibo:    {len(df_user_weibo):>10,}")


✅ user_weibo 去重后:    598,278
✅ user_info.user_id 类型统一为: int64

📦 原始数据量:
  topic_weibo:        4,747
  topic_comment:    121,807
  user_info:         11,012
  user_weibo:       598,278


## Step 2: 文本清洗

清洗原则：
- **保留情绪信号**：表情 `[xxx]` 保留、短文本中的情绪词保留
- **清理噪声**：HTML 标签、URL、多余空白
- **话题标签**：`#xxx#` → 提取内容保留（标签本身含情绪/事件信息）
- **回复前缀**：`回复@xxx：` → 移除前缀但保留正文

In [11]:
def clean_weibo_text(text: str) -> str:
    """清洗微博文本，保留情绪信号。

    清洗步骤（有序）：
    1. 移除 HTML 标签
    2. 移除 URL
    3. 移除 回复@xxx: 前缀（保留正文部分）
    4. 提取话题标签内容（去掉 # 符号，保留文字）
    5. 规范化空白字符
    """
    if not isinstance(text, str) or len(text) == 0:
        return text

    # 1. 移除 HTML 标签
    text = re.sub(r'<[^>]+>', '', text)

    # 2. 移除 URL
    text = re.sub(r'https?://\S+', '', text)

    # 3. 移除 "回复@xxx：" 或 "回复@xxx:" 前缀，保留后续正文
    text = re.sub(r'^回复@[\w\u4e00-\u9fff]+[：:]', '', text)

    # 4. 话题标签：#xxx# → xxx（保留标签文字内容）
    text = re.sub(r'#([^#]+)#', r'\1', text)

    # 5. 规范化空白（多个空白合并为一个，去首尾空白）
    text = re.sub(r'\s+', ' ', text).strip()

    return text


# ========== 应用文本清洗 ==========
print("🧹 正在清洗文本...")

# 保存原始文本备份（用于对比验证）
df_topic_weibo["content_raw"] = df_topic_weibo["content"]
df_topic_comment["content_raw"] = df_topic_comment["content"]
df_user_weibo["content_raw"] = df_user_weibo["content"]

# 应用清洗
df_topic_weibo["content"] = df_topic_weibo["content"].apply(clean_weibo_text)
df_topic_comment["content"] = df_topic_comment["content"].apply(clean_weibo_text)
df_user_weibo["content"] = df_user_weibo["content"].apply(clean_weibo_text)

print("✅ 文本清洗完成")

# 验证清洗效果
print("\n--- 清洗前后对比样例 ---")
changed_mask = df_topic_weibo["content"] != df_topic_weibo["content_raw"]
sample = df_topic_weibo[changed_mask].head(3)
for _, row in sample.iterrows():
    print(f"  原始: {row['content_raw'][:80]}...")
    print(f"  清洗: {row['content'][:80]}...")
    print()

🧹 正在清洗文本...
✅ 文本清洗完成

--- 清洗前后对比样例 ---
  原始: 【#习近平九三重要讲话#】九三胜利日，习近平总书记重要讲话金句来了！#中华民族伟大复兴势不可挡##青春拥抱新时代# ...
  清洗: 【习近平九三重要讲话】九三胜利日，习近平总书记重要讲话金句来了！中华民族伟大复兴势不可挡青春拥抱新时代...

  原始: 【#大理苍山走失8岁男童不幸遇难#，涉事夏令营机构负责人正接受调查】近日，一自闭症男孩在大理参加户外独立营进入苍山失联引关注。8月13日19时许，@新京报 记者...
  清洗: 【大理苍山走失8岁男童不幸遇难，涉事夏令营机构负责人正接受调查】近日，一自闭症男孩在大理参加户外独立营进入苍山失联引关注。8月13日19时许，@新京报 记者从当...

  原始: #张云雷0111生日快乐##张云雷一岁一礼#2023.01.02广州佛山流行金曲嘉年华，主办给张云雷现场过生日，播放粉丝祝张云雷生日快乐的VCR，现场吹蜡烛许愿...
  清洗: 张云雷0111生日快乐张云雷一岁一礼2023.01.02广州佛山流行金曲嘉年华，主办给张云雷现场过生日，播放粉丝祝张云雷生日快乐的VCR，现场吹蜡烛许愿切蛋糕，...



## Step 3: 空文本过滤 & 转发标记

- 移除完全空文本的评论和微博（无法用于情绪分析）
- 对 `user_weibo` 添加 `is_repost` 转发标记列（保留转发关系作为社会行为信号）
- 对内容仅为"转发微博"的帖子，保留记录但标记为无文本价值

In [12]:
system_patterns = [
    "此微博已被作者删除",
    "微博可见时间范围",
    "没有这条微博的查看权限",
    "账号因违反相关法律法规",
    "该微博因违反法律法规",
    "被权利方投诉侵权",
    "用户自行申请关闭", 
    "微博社区公约"
]

ad_keywords = [
        "点开红包", "现金红包", "好礼", "我在参与", 
        "连续签到", "粉打卡", "年度歌曲","免费围观", 
        "关注超话", "森林驿站", "开放公测", "上闲鱼", 
        "头像挂件", "抓马福", "微博智搜", "微博抓马", 
        "马年接福", "春节AI合拍", "微博渔场", "解锁赛博年味", 
        "年度报告", "旅行青蛙中国", "SVIP", "微博之夜", 
        "微博会员", "微博红包"
]

In [13]:
# ========== 3.1 添加转发标记（在过滤前添加，保留行为信息）==========
df_user_weibo["is_repost"] = df_user_weibo["reposted_weibo_id"] != -1
print(f"✅ user_weibo 添加 is_repost 列: 转发 {df_user_weibo['is_repost'].sum():,} 条, "
      f"原创 {(df_user_weibo['is_repost'] == 0).sum():,} 条")

# ========== 3.2 移除空文本记录 ==========
# topic_comment: 空文本评论不再移除，保留并在 3.4 中标记为 Level 0
print(f"✅ topic_comment 保留空文本（将在 3.4 中标记为 Level 0），共 "
      f"{(df_topic_comment['content'].str.strip().str.len() == 0).sum():,} 条")

# user_weibo: 移除空文本微博（但保留"转发微博"，因为它携带转发关系信息）
n_before = len(df_user_weibo)
df_user_weibo = df_user_weibo[df_user_weibo["content"].str.len() > 0].reset_index(drop=True)
print(f"✅ user_weibo 移除空文本: {n_before:,} → {len(df_user_weibo):,} "
      f"(移除 {n_before - len(df_user_weibo):,} 条) | 保留率: {retention_rate('user_weibo', df_user_weibo)}")

# ========== 3.2b 过滤 user_weibo 低价值文本（系统提示 & 广告）==========
# 系统提示：平台生成的无情绪内容（删帖提示、权限限制等）
_system_pattern = "|".join(re.escape(p) for p in system_patterns)
system_mask = df_user_weibo["content"].str.contains(_system_pattern, regex=True, na=False)
n_before = len(df_user_weibo)
df_user_weibo = df_user_weibo[~system_mask].reset_index(drop=True)
print(f"✅ user_weibo 过滤系统提示文本: {n_before:,} → {len(df_user_weibo):,} "
      f"(移除 {n_before - len(df_user_weibo):,} 条) | 保留率: {retention_rate('user_weibo', df_user_weibo)}")

# 广告关键词：平台活动/商业推广，与情绪传播无关
_ad_pattern = "|".join(re.escape(k) for k in ad_keywords)
ad_mask = df_user_weibo["content"].str.contains(_ad_pattern, regex=True, na=False)
n_before = len(df_user_weibo)
df_user_weibo = df_user_weibo[~ad_mask].reset_index(drop=True)
print(f"✅ user_weibo 过滤广告文本: {n_before:,} → {len(df_user_weibo):,} "
      f"(移除 {n_before - len(df_user_weibo):,} 条) | 保留率: {retention_rate('user_weibo', df_user_weibo)}")

# ========== 3.3 低信息量文本判断函数（user_weibo has_text_content 用）==========
def is_low_info_text(text: str) -> bool:
    """判断文本是否为低信息量（纯数字/纯符号/纯英文字母/纯Emoji）。"""
    if not isinstance(text, str) or len(text.strip()) == 0:
        return True
    t = text.strip()
    # 纯数字
    if re.fullmatch(r'\d+', t):
        return True
    # 纯符号（不含字母、数字、汉字、Emoji）
    if re.fullmatch(r'[^\w\u4e00-\u9fff\U00010000-\U0010FFFF]+', t):
        return True
    # 纯英文字母
    if re.fullmatch(r'[a-zA-Z]+', t):
        return True
    # 纯 Emoji
    if re.fullmatch(
        r'[\U0001F300-\U0001F9FF\U00002600-\U000027BF'
        r'\U0001FA00-\U0001FA6F\U0001FA70-\U0001FAFF\uFE00-\uFE0F\u200D]+',
        t
    ):
        return True
    return False

# ========== 3.4 user_weibo has_text_content（二值标记）==========
template_mask = df_user_weibo["content"].isin(["转发微博", "图片评论", "转发"])
low_info_mask_weibo = df_user_weibo["content"].apply(is_low_info_text)
df_user_weibo["has_text_content"] = ~(template_mask | low_info_mask_weibo)
no_text_count_weibo = (~df_user_weibo["has_text_content"]).sum()
print(f"✅ user_weibo 标记 has_text_content: "
      f"有实质文本 {df_user_weibo['has_text_content'].sum():,} 条, "
      f"低信息量 {no_text_count_weibo:,} 条")

# ========== 3.5 topic_comment 文本质量五级标注 ==========
# Level 0: 空文本（空字符串 / 纯空白 / 换行）
_L0_EMPTY = re.compile(r'^\s*$')

# Level 1: 信息量极低（模板转发、纯重复字符、纯数字/符号/Emoji/英文、纯@用户）
_L1_TEMPLATE = re.compile(
    r'^([转轉][发發]微博|图片评论( 评论配图)?|评论配图|哈+|啊+)$'
)
_L1_DIGITS   = re.compile(r'^\d+$')
_L1_SYMBOLS  = re.compile(r'^[^\w\u4e00-\u9fff\U00010000-\U0010FFFF]+$')
_L1_ALPHA    = re.compile(r'^[a-zA-Z]+$')
_L1_EMOJI    = re.compile(
    r'^[\U0001F300-\U0001F9FF\U00002600-\U000027BF'
    r'\U0001FA00-\U0001FA6F\U0001FA70-\U0001FAFF\uFE00-\uFE0F\u200D]+$'
)
_L1_AT_ONLY  = re.compile(r'^(@[\u4e00-\u9fa5a-zA-Z0-9_-]+\s*)+$') 

# Level 2: 有基本语义但情绪分析价值低（寒暄、礼貌互动等）
_L2_GREET = re.compile(
    r'^(感谢分享(精彩内容)?|谢谢(分享|你的支持)?|晚安(网页链接)?'
    r'|早(安|上好([哦呀啊哇])?)?'
    r'|([上中下]午好)([哦呀啊哇])?'
    r'|晚(安|上好([哦呀啊哇])?)?'
    r'|新年快乐|关注|持续关注'
    r'|周(一|二|三|四|五|六|日|末)愉快'
    r'|了解(一下|了)?|签到|确认签收|接(好运|接接)?'
    r'|是的(呢)?|是(啊|呀|的呢?|这样的|的)|对(的|啊)?|说的对'
    r'|没毛病|有道理|嗯(嗯|呢|呐)?|我也觉得|我也是'
    r'|就是(就是|啊|的)?|确实(是这样)?|不错(不错)?|可以'
    r'|赞(同)?|顶|up|哇|啊|哎|来了|好的|看看|好家伙'
    r'|分享知识传播正能量)$',
    re.IGNORECASE
)


def assign_text_quality(text: str) -> int:
    """为 topic_comment 的 content 字段分配文本质量等级（0-4）。

    Returns:
        `int`:
            文本质量等级：
            - 0: 空文本
            - 1: 信息量极低（模板/纯数字/纯符号/纯Emoji/纯英文）
            - 2: 低分析价值（寒暄/礼貌互动）
            - 3: 可分析（默认，有语义和态度线索）
            - 4: 高分析价值（由后续长度阈值提升，此处不处理）
    """
    if not isinstance(text, str):
        return 0
    t = text.strip()
    if _L0_EMPTY.fullmatch(t):
        return 0
    if (
        _L1_TEMPLATE.fullmatch(t)
        or _L1_DIGITS.fullmatch(t)
        or _L1_SYMBOLS.fullmatch(t)
        or _L1_ALPHA.fullmatch(t)
        or _L1_EMOJI.fullmatch(t)
        or _L1_AT_ONLY.fullmatch(t)
    ):
        return 1
    if _L2_GREET.fullmatch(t):
        return 2
    return 3


_QUALITY_LABELS = {0: "空文本", 1: "极低信息", 2: "低分析价值", 3: "可分析", 4: "高分析价值"}

df_topic_comment["text_quality"] = df_topic_comment["content"].apply(assign_text_quality)
df_topic_comment["text_quality_label"] = df_topic_comment["text_quality"].map(_QUALITY_LABELS)

# 打印各等级分布
print(f"\n✅ topic_comment 文本质量标注完成:")
quality_counts = df_topic_comment["text_quality"].value_counts().sort_index()
for level, count in quality_counts.items():
    label = _QUALITY_LABELS[level]
    print(f"  Level {level} ({label}): {count:,} 条 ({count / len(df_topic_comment) * 100:.2f}%)")

print(f"\n📊 当前数据量:")
print(f"  topic_weibo:   {len(df_topic_weibo):>10,}")
print(f"  topic_comment: {len(df_topic_comment):>10,}")
print(f"  user_info:     {len(df_user_info):>10,}")
print(f"  user_weibo:    {len(df_user_weibo):>10,}")

✅ user_weibo 添加 is_repost 列: 转发 220,126 条, 原创 378,152 条
✅ topic_comment 保留空文本（将在 3.4 中标记为 Level 0），共 6,812 条
✅ user_weibo 移除空文本: 598,278 → 590,023 (移除 8,255 条) | 保留率: 590,023 / 598,278 (98.6% 保留)
✅ user_weibo 过滤系统提示文本: 590,023 → 582,359 (移除 7,664 条) | 保留率: 582,359 / 598,278 (97.3% 保留)
✅ user_weibo 过滤广告文本: 582,359 → 567,005 (移除 15,354 条) | 保留率: 567,005 / 598,278 (94.8% 保留)
✅ user_weibo 标记 has_text_content: 有实质文本 489,657 条, 低信息量 77,348 条

✅ topic_comment 文本质量标注完成:
  Level 0 (空文本): 6,812 条 (5.59%)
  Level 1 (极低信息): 3,490 条 (2.87%)
  Level 2 (低分析价值): 3,580 条 (2.94%)
  Level 3 (可分析): 107,925 条 (88.60%)

📊 当前数据量:
  topic_weibo:        4,747
  topic_comment:    121,807
  user_info:         11,012
  user_weibo:       567,005


## Step 4: 特征工程

为后续情绪分析和多智能体行为建模添加关键特征：
- **时间特征**：年/月/日/时/星期
- **文本长度**：用于区分长文讨论和短文情绪表达
- **互动量**：`engagement = like_count + comment_count + repost_count`（衡量信息传播力）
- **用户活跃度**：每位用户的历史微博数量（用于 Agent 行为模式建模）

In [14]:
from pandas import DataFrame

# ========== 4.1 时间特征 ==========
def add_time_features(df: DataFrame, time_col: str = "create_time") -> DataFrame:
    """从时间列提取多维度时间特征。"""
    df["year"] = df[time_col].dt.year
    df["month"] = df[time_col].dt.month
    df["day"] = df[time_col].dt.day
    df["hour"] = df[time_col].dt.hour
    df["weekday"] = df[time_col].dt.day_name()
    return df

df_topic_weibo = add_time_features(df_topic_weibo)
df_topic_comment = add_time_features(df_topic_comment)
df_user_weibo = add_time_features(df_user_weibo)
print("✅ 时间特征添加完成")

# ========== 4.2 文本长度 ==========
df_topic_weibo["text_length"] = df_topic_weibo["content"].str.len()
df_topic_comment["text_length"] = df_topic_comment["content"].str.len()
df_user_weibo["text_length"] = df_user_weibo["content"].str.len()
print("✅ 文本长度特征添加完成")

# ========== 4.3 爬取评论数 ==========
# 统计 topic_comment 中每条 weibo_id 对应的实际爬取评论数（清洗后）
crawled_comment_counts = (
    df_topic_comment.groupby("weibo_id")
    .size()
    .reset_index(name="crawled_comment_count")
)
df_topic_weibo = df_topic_weibo.merge(crawled_comment_counts, on="weibo_id", how="left")
df_topic_weibo["crawled_comment_count"] = df_topic_weibo["crawled_comment_count"].fillna(0).astype(int)
print(f"✅ topic_weibo 添加 crawled_comment_count: "
      f"有评论 {(df_topic_weibo['crawled_comment_count'] > 0).sum()} 条, "
      f"无评论 {(df_topic_weibo['crawled_comment_count'] == 0).sum()} 条")

# 过滤掉未爬取到评论的微博（无法参与情绪传播分析）
n_before = len(df_topic_weibo)
df_topic_weibo = df_topic_weibo[df_topic_weibo["crawled_comment_count"] > 0].reset_index(drop=True)
print(f"✅ topic_weibo 过滤无评论记录: {n_before:,} → {len(df_topic_weibo):,} "
      f"(移除 {n_before - len(df_topic_weibo):,} 条) | 保留率: {retention_rate('topic_weibo', df_topic_weibo)}")

# ========== 4.4 互动量 (engagement) ==========
# topic_weibo: like + comment + repost
df_topic_weibo["engagement"] = (
    df_topic_weibo["like_count"]
    + df_topic_weibo["comment_count"]
    + df_topic_weibo["repost_count"]
)

# topic_comment: like + sub_comment_count
df_topic_comment["engagement"] = (
    df_topic_comment["like_count"]
    + df_topic_comment["sub_comment_count"]
)

# user_weibo: like + comment + repost
df_user_weibo["engagement"] = (
    df_user_weibo["like_count"]
    + df_user_weibo["comment_count"]
    + df_user_weibo["repost_count"]
)
print("✅ 互动量特征添加完成")

# ========== 4.4 用户活跃度统计 ==========
# 计算每位用户在 user_weibo 中的历史微博数量
user_post_counts = df_user_weibo.groupby("user_id").size().reset_index(name="user_post_count")
df_user_weibo = df_user_weibo.merge(user_post_counts, on="user_id", how="left")

# 计算每位用户的原创微博比例
user_original_ratio = (
    df_user_weibo.groupby("user_id")["is_repost"]
    .apply(lambda x: 1 - x.mean())
    .reset_index(name="original_ratio")
)
df_user_weibo = df_user_weibo.merge(user_original_ratio, on="user_id", how="left")
print("✅ 用户活跃度特征添加完成")

# 为 user_info 添加用户活跃度
user_activity = df_user_weibo.groupby("user_id").agg(
    crawled_weibo_count=("weibo_id", "nunique"),
    avg_engagement=("engagement", "mean"),
    original_ratio=("is_repost", lambda x: 1 - x.mean()),
).reset_index()
df_user_info = df_user_info.merge(user_activity, on="user_id", how="left")
# 填充没有微博记录的用户
df_user_info["crawled_weibo_count"] = df_user_info["crawled_weibo_count"].fillna(0).astype(int)
df_user_info["avg_engagement"] = df_user_info["avg_engagement"].fillna(0)
df_user_info["original_ratio"] = df_user_info["original_ratio"].fillna(0)
print("✅ user_info 用户活跃度特征添加完成")

✅ 时间特征添加完成
✅ 文本长度特征添加完成
✅ topic_weibo 添加 crawled_comment_count: 有评论 4704 条, 无评论 43 条
✅ topic_weibo 过滤无评论记录: 4,747 → 4,704 (移除 43 条) | 保留率: 4,704 / 4,747 (99.1% 保留)
✅ 互动量特征添加完成
✅ 用户活跃度特征添加完成
✅ user_info 用户活跃度特征添加完成


## Step 5: user_info 表清洗

- 规范化 `registration_time` 为 datetime 类型
- 清洗 `ip_location` 字段（统一"未知"标记）
- 清洗 `description` 个人简介文本
- 计算账号年龄等辅助特征

In [15]:
# ========== 5.1 registration_time 转 datetime ==========
df_user_info["registration_time"] = pd.to_datetime(df_user_info["registration_time"], errors="coerce")
print(f"✅ registration_time 转为 datetime, 无效值: {df_user_info['registration_time'].isna().sum()}")

# ========== 5.2 计算账号年龄（天数）==========
reference_date = pd.Timestamp("2026-03-01")  # 使用数据采集参考日期
df_user_info["account_age_days"] = (reference_date - df_user_info["registration_time"]).dt.days
print(f"✅ 账号年龄计算完成, 平均: {df_user_info['account_age_days'].mean():.0f} 天")

# ========== 5.3 ip_location 清洗 ==========
# 查看分布
print(f"\n--- IP location 分布 (Top 10) ---")
print(df_user_info["ip_location"].value_counts().head(10))
# "未知" 统一处理
df_user_info["ip_location"] = df_user_info["ip_location"].replace({"未知": None, "": None})
print(f"ip_location 缺失: {df_user_info['ip_location'].isna().sum()}")

# ========== 5.4 description 清洗 ==========
df_user_info["description"] = df_user_info["description"].apply(
    lambda x: clean_weibo_text(x) if isinstance(x, str) else x
)

# ========== 5.5 粉丝/关注比（影响力指标）==========
df_user_info["follower_following_ratio"] = (
    df_user_info["follower_count"] / df_user_info["following_count"].replace(0, 1)
).round(2)
print(f"✅ 粉丝关注比计算完成")

# ========== 5.6 verified 类型整理 ==========
print(f"\n--- 认证类型分布 ---")
print(df_user_info["verified_type_name"].value_counts())

# 查看清洗后的 user_info 结构
print(f"\n✅ user_info 最终列: {df_user_info.columns.tolist()}")
print(f"   Shape: {df_user_info.shape}")

✅ registration_time 转为 datetime, 无效值: 148
✅ 账号年龄计算完成, 平均: 3707 天

--- IP location 分布 (Top 10) ---
ip_location
广东    1465
未知     834
浙江     744
江苏     725
北京     680
山东     656
四川     545
上海     504
河南     446
河北     365
Name: count, dtype: int64
ip_location 缺失: 834
✅ 粉丝关注比计算完成

--- 认证类型分布 ---
verified_type_name
普通用户     7054
个人认证     3806
政府         62
媒体         46
企业         29
团体/机构      10
未知          4
校园          1
Name: count, dtype: int64

✅ user_info 最终列: ['user_id', 'screen_name', 'gender', 'ip_location', 'registration_time', 'total_weibo_count', 'follower_count', 'following_count', 'description', 'verified', 'verified_type', 'verified_type_name', 'user_rank', 'crawled_weibo_count', 'avg_engagement', 'original_ratio', 'account_age_days', 'follower_following_ratio']
   Shape: (11012, 18)


## Step 6: 最终字段整理 & 列顺序规范化

清理临时列（`content_raw`、`create_time_ts`），规范化列顺序，确保数据结构清晰。

In [16]:
# ========== 6.1 topic_weibo 最终字段 ==========
topic_weibo_cols = [
    # ID & 用户
    "weibo_id", "user_id", "screen_name", "gender",
    # 话题
    "topic",
    # 文本
    "content", "text_length",
    # 时间
    "create_time", "year", "month", "day", "hour", "weekday",
    # 互动
    "like_count", "comment_count", "repost_count", "engagement",
    # 爬取评论数
    "crawled_comment_count",
]
df_topic_weibo = df_topic_weibo[topic_weibo_cols]

# ========== 6.2 topic_comment 最终字段 ==========
topic_comment_cols = [
    # ID & 关联
    "comment_id", "weibo_id", "parent_id",
    # 用户
    "user_id", "screen_name", "gender",
    # 文本
    "content", "text_length", "text_quality", "text_quality_label",
    # 时间
    "create_time", "year", "month", "day", "hour", "weekday",
    # 互动
    "like_count", "sub_comment_count", "engagement",
    # 位置
    "ip_location",
]
df_topic_comment = df_topic_comment[topic_comment_cols]

# ========== 6.3 user_weibo 最终字段 ==========
user_weibo_cols = [
    # ID & 用户
    "weibo_id", "user_id", "screen_name",
    # 文本
    "content", "text_length", "has_text_content",
    # 时间
    "create_time", "year", "month", "day", "hour", "weekday",
    # 互动
    "like_count", "comment_count", "repost_count", "engagement",
    # 转发关系
    "is_repost", "reposted_weibo_id",
    # 社交元数据
    "topics", "at_users",
    # 用户维度统计
    "user_post_count", "original_ratio",
]
df_user_weibo = df_user_weibo[user_weibo_cols]

# ========== 6.4 user_info 最终字段 ==========
user_info_cols = [
    # ID
    "user_id", "screen_name", "gender",
    # 位置
    "ip_location",
    # 账号信息
    "registration_time", "account_age_days",
    "verified", "verified_type", "verified_type_name",
    # 社交指标
    "total_weibo_count", "follower_count", "following_count",
    "follower_following_ratio", "user_rank",
    # 活跃度
    "crawled_weibo_count", "avg_engagement", "original_ratio",
    # 个人简介
    "description",
]
df_user_info = df_user_info[user_info_cols]

print("✅ 所有表字段整理完成")
print(f"\n📋 最终数据结构:")
for name, df in [("topic_weibo", df_topic_weibo), ("topic_comment", df_topic_comment),
                  ("user_info", df_user_info), ("user_weibo", df_user_weibo)]:
    print(f"\n  {name} ({df.shape[0]:,} rows × {df.shape[1]} cols)")
    print(f"    {df.columns.tolist()}")

✅ 所有表字段整理完成

📋 最终数据结构:

  topic_weibo (4,704 rows × 18 cols)
    ['weibo_id', 'user_id', 'screen_name', 'gender', 'topic', 'content', 'text_length', 'create_time', 'year', 'month', 'day', 'hour', 'weekday', 'like_count', 'comment_count', 'repost_count', 'engagement', 'crawled_comment_count']

  topic_comment (121,807 rows × 20 cols)
    ['comment_id', 'weibo_id', 'parent_id', 'user_id', 'screen_name', 'gender', 'content', 'text_length', 'text_quality', 'text_quality_label', 'create_time', 'year', 'month', 'day', 'hour', 'weekday', 'like_count', 'sub_comment_count', 'engagement', 'ip_location']

  user_info (11,012 rows × 18 cols)
    ['user_id', 'screen_name', 'gender', 'ip_location', 'registration_time', 'account_age_days', 'verified', 'verified_type', 'verified_type_name', 'total_weibo_count', 'follower_count', 'following_count', 'follower_following_ratio', 'user_rank', 'crawled_weibo_count', 'avg_engagement', 'original_ratio', 'description']

  user_weibo (567,005 rows × 22 cols)
  

## Step 7: 数据质量验证

对清洗后的数据进行全面质量检查，确保：
1. 无重复主键
2. 关键字段无空值
3. 数值字段范围合理
4. 表间关联完整

In [17]:
print("=" * 60)
print("✅ 数据质量验证报告")
print("=" * 60)

all_passed = True

# 1. 主键唯一性
print("\n【1. 主键唯一性】")
checks = [
    ("topic_weibo.weibo_id", df_topic_weibo["weibo_id"].duplicated().sum()),
    ("topic_comment.comment_id", df_topic_comment["comment_id"].duplicated().sum()),
    ("user_info.user_id", df_user_info["user_id"].duplicated().sum()),
]
for name, dup_count in checks:
    status = "✅ PASS" if dup_count == 0 else f"❌ FAIL ({dup_count} duplicates)"
    if dup_count > 0:
        all_passed = False
    print(f"  {name}: {status}")

# 2. 关键字段非空
print("\n【2. 关键字段非空】")
critical_checks = [
    ("topic_weibo.content", df_topic_weibo["content"].isna().sum() + (df_topic_weibo["content"].str.len() == 0).sum()),
    ("topic_comment.content", df_topic_comment["content"].isna().sum() + (df_topic_comment["content"].str.len() == 0).sum()),
    ("user_info.user_id", df_user_info["user_id"].isna().sum()),
    ("user_weibo.content", df_user_weibo["content"].isna().sum()),
]
for name, null_count in critical_checks:
    status = "✅ PASS" if null_count == 0 else f"⚠️ WARN ({null_count} null/empty)"
    if null_count > 0 and "content" not in name:
        all_passed = False
    print(f"  {name}: {status}")

# 3. 数值字段范围
print("\n【3. 数值字段合理性】")
for name, df, cols in [
    ("topic_weibo", df_topic_weibo, ["like_count", "comment_count", "repost_count"]),
    ("topic_comment", df_topic_comment, ["like_count", "sub_comment_count"]),
    ("user_weibo", df_user_weibo, ["like_count", "comment_count", "repost_count"]),
]:
    for col in cols:
        neg_count = (df[col] < 0).sum()
        status = "✅" if neg_count == 0 else f"❌ ({neg_count} negative)"
        if neg_count > 0:
            all_passed = False
        print(f"  {name}.{col}: {status}")

# 4. 表间关联
print("\n【4. 表间关联完整性】")
comment_in_weibo = df_topic_comment["weibo_id"].isin(df_topic_weibo["weibo_id"]).mean()
print(f"  topic_comment → topic_weibo 关联率: {comment_in_weibo:.1%}")

user_weibo_in_info = df_user_weibo["user_id"].isin(df_user_info["user_id"]).mean()
print(f"  user_weibo → user_info 关联率: {user_weibo_in_info:.1%}")

# 5. 时间范围
print("\n【5. 时间范围】")
for name, df in [("topic_weibo", df_topic_weibo), ("topic_comment", df_topic_comment),
                  ("user_weibo", df_user_weibo)]:
    print(f"  {name}: {df['create_time'].min()} ~ {df['create_time'].max()}")

print(f"\n{'='*60}")
print(f"{'✅ 全部验证通过!' if all_passed else '⚠️ 存在需要关注的问题，请查看上方详情'}")
print(f"{'='*60}")

✅ 数据质量验证报告

【1. 主键唯一性】
  topic_weibo.weibo_id: ✅ PASS
  topic_comment.comment_id: ✅ PASS
  user_info.user_id: ✅ PASS

【2. 关键字段非空】
  topic_weibo.content: ✅ PASS
  topic_comment.content: ⚠️ WARN (6812 null/empty)
  user_info.user_id: ✅ PASS
  user_weibo.content: ✅ PASS

【3. 数值字段合理性】
  topic_weibo.like_count: ✅
  topic_weibo.comment_count: ✅
  topic_weibo.repost_count: ✅
  topic_comment.like_count: ✅
  topic_comment.sub_comment_count: ✅
  user_weibo.like_count: ✅
  user_weibo.comment_count: ✅
  user_weibo.repost_count: ✅

【4. 表间关联完整性】
  topic_comment → topic_weibo 关联率: 100.0%
  user_weibo → user_info 关联率: 71.7%

【5. 时间范围】
  topic_weibo: 2025-01-03 11:30:34 ~ 2025-12-31 23:11:13
  topic_comment: 2025-01-03 19:13:02 ~ 2026-02-23 23:50:22
  user_weibo: 2010-07-23 01:39:51 ~ 2026-03-08 10:57:50

✅ 全部验证通过!


## Step 8: 保存清洗后的数据

将清洗后的四个数据表保存为 Parquet 格式（高效压缩，保留类型信息），存放至 `data/cleaned/` 目录。

In [18]:
import os

# 创建输出目录
output_dir = r"..\data\cleaned"
os.makedirs(output_dir, exist_ok=True)

# 保存
datasets = {
    "topic_weibo": df_topic_weibo,
    "topic_comment": df_topic_comment,
    "user_info": df_user_info,
    "user_weibo": df_user_weibo,
}

for name, df in datasets.items():
    path = os.path.join(output_dir, f"{name}.parquet")
    df.to_parquet(path, index=False)
    size_mb = os.path.getsize(path) / (1024 * 1024)
    print(f"✅ {name}.parquet 已保存 ({df.shape[0]:>10,} rows × {df.shape[1]:>2} cols, {size_mb:.1f} MB)")

print(f"\n📂 输出目录: {os.path.abspath(output_dir)}")

✅ topic_weibo.parquet 已保存 (     4,704 rows × 18 cols, 2.0 MB)
✅ topic_comment.parquet 已保存 (   121,807 rows × 20 cols, 10.0 MB)
✅ user_info.parquet 已保存 (    11,012 rows × 18 cols, 0.9 MB)
✅ user_weibo.parquet 已保存 (   567,005 rows × 22 cols, 131.6 MB)

📂 输出目录: d:\GraduationProject\data\cleaned


## 📋 数据清洗总结

### 清洗操作汇总

| 步骤 | 操作 | 影响 |
|------|------|------|
| 去重 | `user_weibo` 按 (weibo_id, user_id) 去重 | 919,646 → 598,278 行 |
| 类型统一 | `user_info.user_id` object → int64 | 确保表间关联 |
| 文本清洗 | 移除 HTML/URL，提取话题标签内容，清理回复前缀 | 保留情绪信号 |
| 空文本 | `user_weibo` 移除空文本；`topic_comment` 保留并标记 Level 0 | 提升文本质量 |
| 低价值过滤 | `user_weibo` 过滤系统提示、广告关键词文本 | 剔除平台噪声 |
| 转发标记 | 添加 `is_repost`、`has_text_content` 列 | 保留行为信号 |
| 文本质量分级 | `topic_comment` 添加 `text_quality`（0-4 级）和 `text_quality_label` | 支持精细化情绪分析 |
| 时间特征 | 提取 year/month/day/hour/weekday | 支持时序分析 |
| 互动量 | `engagement` = 点赞+评论+转发 | 衡量传播力 |
| 用户特征 | 账号年龄、粉丝关注比、活跃度、原创比例 | Agent 行为建模 |

### 最终数据规模

| 数据表 | 行数 | 列数 | 说明 |
|--------|------|------|------|
| `topic_weibo` | 4,747 | 17 | 热点话题微博 |
| `topic_comment` | 114,995 | 20 | 话题评论（含 5 级文本质量标注） |
| `user_info` | 11,012 | 18 | 用户画像（含活跃度/影响力特征） |
| `user_weibo` | 590,023 | 22 | 用户历史微博（含转发关系/社交元数据） |

### 保留的社会行为信号
- ✅ **转发关系**：`reposted_weibo_id` + `is_repost` 标记
- ✅ **评论层级**：`parent_id` 支持评论树重构
- ✅ **@互动**：`at_users` 字段保留用户间提及关系
- ✅ **话题参与**：`topics` 字段保留用户关注的话题
- ✅ **微博表情**：`[xxx]` 格式的表情符号保留，作为情绪信号
- ✅ **互动量**：engagement 综合衡量信息传播影响力
- ✅ **文本质量分级**：`text_quality` 0-4 级标注，`text_quality_label` 中文描述

### 文本质量等级说明（topic_comment）

| 等级 | 标签 | 描述 | 示例 |
|------|------|------|------|
| 0 | 空文本 | 空字符串、纯空白、换行 | `""`, `" "` |
| 1 | 极低信息 | 模板转发、纯数字/符号/Emoji/英文、重复字符 | `"转发微博"`, `"233"`, `"哈哈哈"` |
| 2 | 低分析价值 | 寒暄、礼貌互动、简单确认 | `"早上好"`, `"感谢分享"`, `"是的"` |
| 3 | 可分析 | 有语义和态度/情绪线索（默认） | 大多数正常评论 |
| 4 | 高分析价值 | 强情感表达、长文讨论（待后续提升） | — |

### 输出路径
`data/cleaned/` 目录下的 4 个 Parquet 文件。

In [19]:
import pandas as pd

df_topic_weibo = pd.read_parquet(r"..\data\cleaned\topic_weibo.parquet")
df_topic_comment = pd.read_parquet(r"..\data\cleaned\topic_comment.parquet")
df_user_info = pd.read_parquet(r"..\data\cleaned\user_info.parquet")
df_user_weibo = pd.read_parquet(r"..\data\cleaned\user_weibo.parquet")

In [20]:
len(df_user_weibo), len(df_topic_comment)

(567005, 121807)

In [21]:
# df_user_weibo["content"].value_counts()

In [22]:
# df_topic_weibo[["weibo_id", "content"]].merge(
#     df_topic_comment[["weibo_id", "content", "text_quality"]][df_topic_comment["content"].str.contains(r"早安|早上好|晚上好|晚安|下午好|晚安", regex=True)],
#     on="weibo_id",
#     suffixes=("_weibo", "_comment")
# ).query("text_quality > 2")